### PR TITLE
binutils: detect the "gold" and "headers" variants

### LINKED ISSUE #29350: `llvm+gold` with external `binutils` build failure because of missing `plugin-api.h` header
### Steps to reproduce

```
git clone -c feature.manyFiles=true https://github.com/spack/spack.git
spack install llvm@13.0.1
```

### Error message

```
==> Installing llvm-13.0.1-uz4giobhbnsvrlmkq6siezxk6hxedpuw
==> No binary for llvm-13.0.1-uz4giobhbnsvrlmkq6siezxk6hxedpuw found: installing from source
==> Fetching https://mirror.spack.io/_source-cache/archive/09/09c50d558bd975c41157364421820228df66632802a4a6a7c9c17f86a7340802.tar.gz
==> No patches needed for llvm
==> llvm: Executing phase: 'cmake'
==> llvm: Executing phase: 'build'
==> Error: ProcessError: Command exited with status 1:
    'ninja' '-j16' '-v'

2 errors found in build log:
...
  >> 4592    FAILED: tools/gold/CMakeFiles/LLVMgold.dir/gold-plugin.cpp.o
     4593    /home/rutvikc2/spack/lib/spack/env/gcc/g++  -D_GNU_SOURCE -D__STDC_CONSTANT_MACRO
             S -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -Itools/gold -I/tmp/rutvikc2/spack
             -stage/spack-stage-llvm-13.0.1-uz4giobhbnsvrlmkq6siezxk6hxedpuw/spack-src/llvm/to
             ols/gold -Iinclude -I/tmp/rutvikc2/spack-stage/spack-stage-llvm-13.0.1-uz4giobhbn
             svrlmkq6siezxk6hxedpuw/spack-src/llvm/include -isystem /home/rutvikc2/spack/opt/s
             pack/linux-ubuntu20.04-skylake_avx512/gcc-9.3.0/zlib-1.2.11-ulsjxh4r33cuzn6puqa37
             zmpj4n54yyb/include -std=c++11 -fPIC -fno-semantic-interposition -fvisibility-inl
             ines-hidden -Werror=date-time -Wall -Wextra -Wno-unused-parameter -Wwrite-strings
              -Wcast-qual -Wno-missing-field-initializers -pedantic -Wno-long-long -Wimplicit-
             fallthrough -Wno-maybe-uninitialized -Wno-class-memaccess -Wno-redundant-move -Wn
             o-pessimizing-move -Wno-noexcept-type -Wdelete-non-virtual-dtor -Wsuggest-overrid
             e -Wno-comment -Wmisleading-indentation -fdiagnostics-color -ffunction-sections -
             fdata-sections -O3 -DNDEBUG -fPIC     -std=c++14 -MD -MT tools/gold/CMakeFiles/LL
             VMgold.dir/gold-plugin.cpp.o -MF tools/gold/CMakeFiles/LLVMgold.dir/gold-plugin.c
             pp.o.d -o tools/gold/CMakeFiles/LLVMgold.dir/gold-plugin.cpp.o -c /tmp/rutvikc2/s
             pack-stage/spack-stage-llvm-13.0.1-uz4giobhbnsvrlmkq6siezxk6hxedpuw/spack-src/llv
             m/tools/gold/gold-plugin.cpp
  >> 4594    /tmp/rutvikc2/spack-stage/spack-stage-llvm-13.0.1-uz4giobhbnsvrlmkq6siezxk6hxedpu
             w/spack-src/llvm/tools/gold/gold-plugin.cpp:38:10: fatal error: plugin-api.h: No 
             such file or directory
     4595       38 | #include <plugin-api.h>
     4596          |          ^~~~~~~~~~~~~~
     4597    compilation terminated.
```

### Information on your system

* **Spack:** 0.17.1-1457-4081597731
* **Python:** 3.8.10
* **Platform:** linux-ubuntu20.04-skylake_avx512
* **Concretizer:** clingo

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have searched the issues of this repo and believe this is not a duplicate
- [X] I have run the failing commands in debug mode and reported the output

### PR DESCRIPTION
**Pull request by bernhardkaindl**
Fixes #29350:

Changes:

Add initial `def determine_variants()`:

- `llvm` needs `binutils+headers` for `plugin-api.h` which most distros don't install
  and therefore make `spack external find binutils` search for the header and set `~headers/+headers`
  
- Also check for installed `gold` linker and set `+gold`/`~gold`

### CODE REVIEW

**Comment by naromero77** [comment]
I have run into this same issue with LLVM. I don't understand the patch though and will wait for @alalazo to provide a proper review.

**Comment by bonachea** [comment]
Ping @alalazo is this ever going to be fixed? 

binutils external detection remains broken in spack:develop, which breaks LLVM and other packages on important systems (e.g. HPE Cray EX, where the binutils provided by the SLES15 distro is missing the `plugin-api.h` header, and possibly other things).

Currently the external detection for `binutils` effectively lies to Spack about what variants are installed, without verifying anything, leading to breakage of packages who need specific binutils variants. We need `binutils` external detection to either be fixed to correctly detect which variants are _actually_ installed, or otherwise disable the external detection in binutils entirely to make it stop lying about what's available.

CC: @arezaii

File: var/spack/repos/builtin/packages/binutils/package.py [hunk 1]
```diff
@@ -168,6 +168,12 @@ def determine_version(cls, exe):
         match = re.search(r"GNU (nm|readelf).* (\S+)", output)
         return Version(match.group(2)).dotted.up_to(3) if match else None
 
+    @classmethod
+    def determine_variants(cls, exes, version_str):
+        variants = "+gold" if "gold" in os.listdir(os.path.dirname(exes[0])) else "~gold"
+        # 29350: llvm needs +headers for plugin-api.h which most distros don't install:
+        return "~headers" + variants
```
Comment (alalazo) [inline var/spack/repos/builtin/packages/binutils/package.py]: We shouldn't blindly return `~headers`, but try instead to check if anything like `binutils-dev` was installed. Also, even if we do so, we won't be able to automatically detect changes in the system.

**Review by alalazo — CHANGES_REQUESTED** [review]

File: var/spack/repos/builtin/packages/binutils/package.py (hunk 1 above)
Comment (bonachea) [reply var/spack/repos/builtin/packages/binutils/package.py]: @alalazo wrote:
> We should ... check if anything like `binutils-dev` was installed. 

I agree that would be the ideal behavior for `determine_variants`. Are you offering to implement that? (I lack the necessary `binutils` expertise)

> Also, even if we do so, we won't be able to automatically detect changes in the system.

IIUC this property is basically true of all spack externals, so that's fundamental. 

Towards your larger point, I agree the implications of unconditionally returning `~headers` is not ideal. IIUC it means that dependent packages who request `binutils+headers` will force concretization that rebuilds `binutils`, which _might_ be unnecessary for the small fraction of systems that have a external install of `binutils` that includes headers (and if for some unrelated reason that build fails, users with a `binutils-dev` package installed still have the option to manually edit the external to add the relevant variant). 

However the status quo of unconditionally (effectively) returning `+headers` seems far worse: dependent packages like LLVM who request `binutils+headers` because they actually need it are instead handed a non-compliant external `binutils`, and then fail to build in messy ways, with the obscure root cause that Spack's auto-detection lied to them about the external's variants. This problem wasted several weeks of developer time before we identified the real root cause; IMHO this should be fixed so others need not go through that exercise.

Comment (bernhardkaindl) [reply var/spack/repos/builtin/packages/binutils/package.py]: @bonachea 👍 💯% agreed! We need to stop wasting user time:

> This problem wasted several weeks of developer time before we identified the real root cause; IMHO this should be fixed so others need not go through that exercise.

Exactly. It is the responsibility of those who what to claim +headers is available to check that it is available. The default has to be `~headers`:

Only 3 packages request binutils.*+headers: llvm, tau, apex. If people building those and gain so much that they would not want to build binutils+headers in spack, it is on those people to do check that the needed headers are available. As long as the found binutils have not proven all headers (not just `plugin-api.h`) are installed and working to build those packages, `~headers` has to be set.

Adding `+headers` is an optional feature for later.
This fix for `~headers` is the failsafe default that we should not even argue against to merge now:

Mind that, compared to the build time that building `llvm` requires, building `binutils` (just in doubt) is not even noticeable because `binutils+headers` builds so quickly compared to the failing `llvm` build.

It's a no-brainer to merge this PR as it.

Given from the many people like @naromero77, @wyphan (header detection in #30954, but that is an incremental later change) with #29350 #31706, and @bonachea with #41024 now, we have enough proven affected people to take action and merge this fix, before we consider how we can detect +headers.

See #30954 for a possible followup after this fix is applied, but merge this first.

Comment (alalazo) [reply var/spack/repos/builtin/packages/binutils/package.py]: @bernhardkaindl Do you have time to add some heuristic check to decide whether to stick `~headers` or `+headers`? I might have time in ~2 weeks. 

> Adding +headers is an optional feature for later.

Meh, I understand this gets past the problem above, but I already see the next issue incoming "Why Spack detects `binutils~headers` when I have headers installed"? 

Right now Spack does the right thing on detection, in the sense that it doesn't add information about headers. 

It does imo the wrong thing during concretization, in the sense that externals _without_ any information on the `headers` variant will be good candidate for _both_ `~headers` and `+headers`. I tried to advocate for the opposite approach (an external with missing information about a variant shouldn't be considered when the variant is required explicitly, no matter the value), with this proof of concept #36179 but I lost the battle.

That said, I agree if we want to add proper detection for the variant, it just shouldn't be hard-coded `~headers`.

Comment (alalazo) [reply var/spack/repos/builtin/packages/binutils/package.py]: @bonachea This is slightly ot, but I'm genuinely curious about your workflow. May I ask why and how you end up with binutils detected? Are you generally preferring to use externals rather than building/reusing binaries, or there was any other motivation for that?

Comment (alalazo) [reply var/spack/repos/builtin/packages/binutils/package.py]: `bfd.h` seems a good representative header file

Comment (bonachea) [reply var/spack/repos/builtin/packages/binutils/package.py]: > @bonachea This is slightly ot, but I'm genuinely curious about your workflow. May I ask why and how you end up with binutils detected? Are you generally preferring to use externals rather than building/reusing binaries, or there was any other motivation for that?

@alalazo: When I'm starting on a new system I always start with `spack external find`, which is [documented here for end-users](https://spack.readthedocs.io/en/latest/packages_yaml.html#automatically-find-external-packages). This is easy to type and locates ALL the detectable system packages, and has the following benefits: 
1. It reduces Spack build times by not rebuilding software you already have
2. It reduces the space footprint on systems with tight disk quotas
3. It's basically mandatory on tightly-integrated HPC systems like Crays where much of the critical software cannot be built correctly by Spack (often because it's closed-source, or the Spack package lacks critical site-specific tweaks)
4. It avoids [security problems documented here](https://spack.readthedocs.io/en/latest/getting_started.html#openssl) with fundamental security packages like `openssl`

The cynical answer to your implied question is "If think it's a bad idea for folks to detect binutils as an external, then you should disable its external detection". IMHO even that "solution" would be greatly preferable to the debacle caused by the current, critically defective detection logic.

I'll also note that adopting this PR as-is today (or simply disabling binutils external detection entirely) doesn't prevent anyone from implementing robust external variant detection _later_. IOW it solves the immediate pressing problem that's impacting many users, until that preferred upgrade is potentially applied later. To put it another way, if your building is on fire, first put out the fire, and maybe later look into installing a sprinkler system.

**Review by bonachea — APPROVED** [review]

File: var/spack/repos/builtin/packages/binutils/package.py (hunk 1 above)
Comment (alalazo) [reply var/spack/repos/builtin/packages/binutils/package.py]: > The cynical answer to your implied question is "If think it's a bad idea for folks to detect binutils as an external, then you should disable its external detection". IMHO even that "solution" would be greatly preferable to the debacle caused by the current, critically defective detection logic.

I didn't imply any question, cynicism is in the eye of the beholder :wink: Also, the current detection isn't wrong in any way - it just doesn't detect `headers`. The [docs](https://spack.readthedocs.io/en/latest/packages_yaml.html#external-packages) say this about externals:

   _Each external should specify a spec: string that should be as well-defined as reasonably possible. If a package lacks a spec component, such as missing a compiler or package version, then Spack will guess the missing component based on its most-favored packages, and it may guess incorrectly._

Tight quota has never been in my thought, so TIL.

> It's basically mandatory on tightly-integrated HPC systems like Crays [ ... ]

We have:
```console
$ spack external read-cray-manifest
```
specifically for Cray. To my understanding, this should give a much more accurate result.

---

**TL;DR** I didn't think I'd generate so much churn, asking to add a few lines of code to make this PR correct and mergeable. So, in the end I installed the sprinkler myself, in https://github.com/spack/spack/pull/40214/commits/8e1ac7ce8b16ab68dc31752ed6e04da5b0e719ee Hope nobody minds.

If you confirm this works I'll merge the PR.

**Review by bonachea — APPROVED** [review]
@alalazo wrote:
> ...
> I installed the sprinkler myself, in [8e1ac7c](https://github.com/spack/spack/commit/8e1ac7ce8b16ab68dc31752ed6e04da5b0e719ee) Hope nobody minds.
> 
> If you confirm this works I'll merge the PR.

Fix confirmed on NERSC Perlmutter:
```
perlmutter$ spack external find binutils
==> The following specs have been detected on this system and added to <redacted>/packages.yaml
binutils@2.40  binutils@2.41.0

perlmutter$ grep -A6 binutils $SPACK_USER_CONFIG_PATH/packages.yaml 
    externals:
    - spec: binutils@2.40~gold~headers
      prefix: /opt/cray/pe/cce/17.0.0/binutils/x86_64/x86_64-pc-linux-gnu
    - spec: binutils@2.41.0~gold~headers
      prefix: /usr
```
Thanks for the quick turnaround!

**Review by alalazo — APPROVED** [review]
